Add game modes.

patrycja93 opened the pull request:
Introduced new abstract class and implementations for it due to the share code of normal games between two players and automatic tests. In the future will be posibility to play with a computer.

Documentation is not updated yet. Changed UIFactory class to return UI depending to the amount of input array parameters.

patrycja93 committed:
Introduced new abstract class and implementations for it due to share code of normal game between two players and automatic tests. In the future will be posibility to play with a computer.
Documentation is not updated yet. Changed UIFactory class to return UI depending to amount of input array parameters.

BulaMar commented on src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java:
```diff
@@ -0,0 +1,82 @@
+
+    private GameSettings setup(String[] inputArrayParameters) {
+        checkCorrectInputData(inputArrayParameters);
+        //TODO: generate draw sequence!!
```
Should this be here? Where are you generating other sequences?

patrycja93 replied:
Yes, good remark. I deleted the method generating sequence but I forgot to change the place for the todo.

BulaMar commented on src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    private GameSettings setup(String[] inputArrayParameters) {
+        checkCorrectInputData(inputArrayParameters);
```
Are you checking if input data is correct or checking correct input data?

patrycja93 replied:
Ok, I will change this name.

BulaMar commented on src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    private GameSettings updateValuesInSettings(String[] inputArrayParameters) {
+        GameSettings.GameSettingsBuilder gameSettings = GameSettings.builder();
```
Could you collect all lines of builder into return statement and build it then?
Also this is not `gameSettings` but `gameSettingsBuilder` - not important when you will apply first line of this comment.

patrycja93 replied:
Ok, I will move it to the return statement.

BulaMar commented on src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java:
```diff
@@ -0,0 +1,82 @@
+        gameSettings.boardSize(maxBoardSize);
+        gameSettings.unbrokenLine(minUnbrokenLine);
+
+        int towards = Integer.parseInt(inputArrayParameters[2]);
```
Please move this line under `int maxBoardSize = Math.max(boardSize, unbrokenLine);` that all `inputArrayParameters` values are in one place.

patrycja93 replied:
Ok, I will do it.

BulaMar commented on src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java:
```diff
@@ -0,0 +1,82 @@
+                return numberOfMatchesDiagonals(boardSize, unbrokenLine);
+            }
+            default:
+                return 3;
```
Please make constants with meaningful names from this cases.
For example: HORIZONTAL = 0

Additionally if you will already have it as constants try change switch case into map with functions.

patrycja93 replied:
Do you mean constants inside the method?

patrycja93 replied:
I made it inside the class as a private final.

BulaMar commented on src/main/java/pl/patrycja/ox/gameflow/FactoryMode.java:
```diff
@@ -0,0 +1,35 @@
+    //create game settings object and set board size
+    //unbroken line, ui taking amount of args
+    Mode setMode() {
+        switch (inputParametersSize) {
```
Try to replace with map. You do not want to adding each time case condition if new mode appear.

patrycja93 replied:
Ok, I will change it.

BulaMar commented on src/main/java/pl/patrycja/ox/gameflow/Game.java:
```diff
@@ -2,31 +2,45 @@
-                    .addController(playerChanger)
-                    .start();
+    private GameSettings updateValuesForBoardAndUnbrokenLine(String[] inputArrayParameters) {
+        GameSettings.GameSettingsBuilder gameSettings = GameSettings.builder();
```
Same situation with builder here.

patrycja93 replied:
ok  I will change it.

BulaMar commented on src/main/java/pl/patrycja/ox/gameflow/Game.java:
```diff
@@ -2,31 +2,45 @@
-                    .start();
+    private GameSettings updateValuesForBoardAndUnbrokenLine(String[] inputArrayParameters) {
+        GameSettings.GameSettingsBuilder gameSettings = GameSettings.builder();
+        int firstParameter = Integer.parseInt(inputArrayParameters[0]);
```
Is that board size, maybe worth to name it like that?

patrycja93 replied:
Yes, you are right.

BulaMar commented on src/main/java/pl/patrycja/ox/gameflow/Game.java:
```diff
@@ -2,31 +2,45 @@
+        if (inputArrayParameters.length == 1) {
+            gameSettings.boardSize(Math.max(firstParameter, 3));
+        } else {
+            int secondParameter = Integer.parseInt(inputArrayParameters[1]);
```
unbroken line?

patrycja93 replied:
I will change it.

BulaMar commented on src/main/java/pl/patrycja/ox/gameflow/Mode.java:
```diff
@@ -2,77 +2,40 @@
-            setBoardSizeAndUnbrokenLine(gameSettings, boardSize, unbrokenLine, towards);
+        for (int i = 0; i < gameSettings.getNumberOfMatches(); i++) {
+            List<Spectator> spectators = SpectatorsRoom.addSpectators(gameSettings);
+            //TODO: changes game setting on boardSize
```
What exactly are you going to do here?

patrycja93 replied:
I will explain to you personally 😄

BulaMar commented on src/main/java/pl/patrycja/ox/ui/InputChecker.java:
```diff
@@ -32,14 +28,14 @@ public int getValidNumber() {
     }
 
-    public boolean getValidNumberDemoMode(String boardSize, String unbrokenLine) {
+    public void getValidNumberFromInputArray(String[] parameters) {
```
Will that still be a `get` type method?

patrycja93 replied:
I will change it.

BulaMar commented on src/main/java/pl/patrycja/ox/ui/InputChecker.java:
```diff
@@ -32,14 +28,14 @@ public int getValidNumber() {
         } catch (IllegalArgumentException e) {
-            e.printStackTrace();
-            return false;
+            ui.display("Entered wrong arguments. Please enter integer numbers and run the game again.");
```
`Please run game again with correct integer numbers`?

patrycja93 replied:
Ok.

BulaMar commented on src/main/java/pl/patrycja/ox/ui/UIFactory.java:
```diff
@@ -1,13 +1,32 @@
+        return ui;
+    }
+
+    private Map<Integer, UI> setUi() {
```
👍  good that this is not a switch this time but do you need 3 times `new ConsoleUI()`

patrycja93 replied:
Probably I do not need it. I will think about it through.

BulaMar reviewed (changes requested):
Generally nice start for extendable modes in future but still need some correction in this state.

patrycja93 committed:
After code review.
Changed switch cases to map with lambdas. Corrected little grammatic errors. Added checking if board size is lower than the unbroken line in GameSetting class. Connected creating GameSetting object.

BulaMar reviewed (changes requested):
Thanks for making all requested changes.
Please fix failing builds.

patrycja93 committed:
Corrected tests.
